### PR TITLE
baggage: Improve performance

### PR DESCRIPTION
**Pull request by cdvr1993**
Several of our microservices during profiling report a considerable cost coming from the regex that validate the key and the value (in the range of 3-6%). For instance:

Screenshot 2023-12-04 at 4 32 35 PM

After checking the regex I saw that we can easily replace that to a simple for loop implementation. All tests are passing, but hopefully didn't miss a condition.

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/baggage
cpu: AMD EPYC 7B13
             │ /tmp/old.txt │            /tmp/new.txt             │
             │    sec/op    │   sec/op     vs base                │
New-96          1.394µ ± 5%   1.382µ ± 7%        ~ (p=0.190 n=10)
NewMember-96   307.10n ± 1%   39.66n ± 1%  -87.08% (p=0.000 n=10)
Parse-96        4.575µ ± 6%   1.037µ ± 7%  -77.33% (p=0.000 n=10)
geomean         1.251µ        384.5n       -69.26%

             │ /tmp/old.txt  │             /tmp/new.txt             │
             │     B/op      │    B/op     vs base                  │
New-96          835.0 ± 0%     835.0 ± 0%        ~ (p=1.000 n=10) ¹
NewMember-96    0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
Parse-96       1108.0 ± 0%     832.0 ± 0%  -24.91% (p=0.000 n=10)
geomean                    ²                -9.11%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

             │ /tmp/old.txt  │             /tmp/new.txt             │
             │   allocs/op   │ allocs/op   vs base                  │
New-96          16.00 ± 0%     16.00 ± 0%        ~ (p=1.000 n=10) ¹
NewMember-96    0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
Parse-96       11.000 ± 0%     7.000 ± 0%  -36.36% (p=0.000 n=10)
geomean                    ²               -13.99%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

### CODE REVIEW

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:45]: How about removing this regex as well?

**Comment by pellared** [comment]
Please fix the build (lint) failures.

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go:45]: This regex is slightly more complex and it is actually used to get subgroups:

```
match := propertyRe.FindStringSubmatch(property)
	if len(match) != 4
```

The ones I changed are simple and only used for validation.

**Comment by pellared** [comment]
Can you please run all benchmarks for the whole package? The changes should also affect `BenchmarkNew` and `BenchmarkParse`.

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:598]: How about refactoring to a simple functions named e.g. `validateKey` and `validateValue`?

File: baggage/baggage.go
Comment (pellared) [reply baggage/baggage.go:45]: This is true. However, I expect it could be a significant improvement for `Parse`. Is it something that you would like to take a look (as a separate PR)?

**Review by MrAlias — COMMENTED** [review]
Thank you the contribution.

I agree with @pellared's comments in that this looks like a partial solution. There is coupling with other functionality that this does not address. I do not think these changes should be accepted as they currently are presented.

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go:558]: Just to reduce the number of changes, but I will update it as you suggest.

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go:43]: Working on it.

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go:548]: sure, didn't want to make even more changes to the code, but we can definitely do that

Comment (pellared) [inline baggage/baggage.go:548]: How about renaming to e.g. `parseProperty`? Can we also make a regular function comment which tells what the function accepts and returns? We can reference the W3C Baggage specification instead of mentioning some regex. I think it would be even good enough to reference https://www.w3.org/TR/baggage in `doc.go`.

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:612]: Can we rename it to `validateKey` or `keyValid`?

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:640]: Can we rename it to `validateValue` or `valueValid`?

**Comment by cdvr1993** [comment]
@pellared what about now?

File: baggage/baggage.go
Comment (pellared) [reply baggage/baggage.go:566]: https://github.com/open-telemetry/opentelemetry-go/pull/4743/commits/00be1c90f8c0ad795a30760514452b1bd941c89d

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:607]: The character validation skips the invalid chars as well as valid leading and trailing whitespaces. 

Then we skip the valid leading and trailing whitespaces.

Then for key part we make `if s[index] != keyValueDelimiter[0]` and for value part we make `index != len(s)` check if there were no invalid chars in the middle of the keo or value.

I find it extremely hard to follow. I spend literally 3 hours to analyze this code and I do not find it maintainable in the shape as it is currently written.

What do you think about the following code? I bet it will have similar performance and I think it would be easier to follow for most people.

```suggestion
func isWhitespace(c rune) bool {
	switch c {
	case ' ', '\t', '\n', '\v', '\f', '\r':
		return true
	default:
		return false
	}
}

func isKeyValueDelimiter(c rune) bool {
	return c == '='
}

// parsePropertyInternal attempts to decode a Property from the passed string.
func parsePropertyInternal(s string) (p Property, ok bool) {
	i := 0

	// Parse key.

	// Skip leading key whitespaces.
	for _, c := range s[i:] {
		if !isWhitespace(c) {
			break
		}
		i++
	}

	keyIdxStart := i
	gotDelimiter := false
	gotWhitespace := false
	for _, c := range s[i:] {
		if isKeyValueDelimiter(c) {
			// We got the delimiter so we can move to the value parsing.
			gotDelimiter = true
			break
		}
		if isWhitespace(c) {
			// We are going to check if there are trailing key whitespaces.
			gotWhitespace = true
			break
		}
		if !validateKeyChar(c) {
			// There is an invalid key character.
			return
		}
		i++
	}

	keyIdxEnd := i

	if keyIdxStart == keyIdxEnd {
		// Empty key is invalid.
		return
	}

	if gotWhitespace {
		for _, c := range s[i:] {
			if isKeyValueDelimiter(c) {
				// We got the delimiter so we can move to the value parsing.
				gotDelimiter = true
				break
			}
			if !isWhitespace(c) {
				// There are other characters than leading whitespaces.
				return
			}
			i++
		}
	}

	if !gotDelimiter {
		ok = true
		p.key = s[keyIdxStart:keyIdxEnd]
		return
	}

	// Skip the seperator.
	i++

	// Parse value.

	// Skip leading value whitespaces.
	for _, c := range s[i:] {
		if !isWhitespace(c) {
			break
		}
		i++
	}

	valueIdxStart := i

	gotWhitespace = false
	for _, c := range s[i:] {
		if isWhitespace(c) {
			// We are going to check if there are trailing value whitespaces.
			gotWhitespace = true
			break
		}
		if !validateValueChar(c) {
			// There is an invalid value character.
			return
		}
		i++
	}

	valueIdxEnd := i

	if gotWhitespace {
		for _, c := range s[i:] {
			if !isWhitespace(c) {
				// There are other characters than leading whitespaces.
				return
			}
			i++
		}
	}

	ok = true
	p.key = s[keyIdxStart:keyIdxEnd]
	p.value = s[valueIdxStart:valueIdxEnd]
	p.hasValue = true
	return
}
```

Comment (pellared) [reply baggage/baggage.go:607]: Remark: I am fine to accept a different implementation than my proposal. I just find the current implementation too complex and brittle. If you would disagree we can ask other approvers for opinions. Still, I hope that you will like my suggestion 😉

Comment (cdvr1993) [reply baggage/baggage.go:607]: I'm fine by using your version. At the end of the day the only thing I care about primarily is about improving the performance about NewMember(). I just tried to the parseProperty because you suggested to do it so that my changes were coupled to that.

After saying that, my personal opinion is that I find your version more complex than mine, probably easier to read, but it has more branches than mine.

My version was only mimicking what we do in the regex, we have the following:

```
^(?:\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*|\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*=\s*([\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*)\s*)$.
```

this can be divided in 2:
```
^\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*$

or

\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*=\s*([\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*)\s*$
```

so the code looks like:

```
func parsePropertyInternal(s string) (p Property, ok bool) {
        // \s*
	index := skipSpace(s, 0)

        // [\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+
	keyStart := index
	keyEnd := index
	for _, c := range s[keyStart:] {
		if !validateKeyChar(c) {
			break
		}
		keyEnd++
	}

        // empty key 
	if keyStart == keyEnd {
		return
	}

        // \s*
	index = skipSpace(s, keyEnd)

	p.key = s[keyStart:keyEnd]

	// this matches only the key
        // first case: ^\s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*$
	if index == len(s) {
		ok = true
		return
	}

        // now let's move to the second case: \s*([\x21\x23-\x27\x2A\x2B\x2D\x2E\x30-\x39\x41-\x5a\x5e-\x7a\x7c\x7e]+)\s*=\s*([\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*)\s*$
        // this has the same pattern for the key, so we just need to look now for '='
	// now let see if it matches the key and the value
	if s[index] != keyValueDelimiter[0] {
                 // doesn't match the expect char, ok is false, so invalid property
		return
	}

	// we set it to true as soon as we find the delimiter even if it is empty.
        // we don't care to set it as true if the value is invalid, because the caller must check first the boolean return value
	p.hasValue = true

        // \s*
	index = skipSpace(s, index+1)

        // [\x21\x23-\x2b\x2d-\x3a\x3c-\x5B\x5D-\x7e]*
	valueStart := index
	valueEnd := index
	for _, c := range s[valueStart:] {
		if !validateValueChar(c) {
			break
		}
		valueEnd++
	}

        // \s*
	index = skipSpace(s, valueEnd)

        // if after checking for trailing whitespaces for the value, we have not finished, it means there are invalid characters, so ok is still false
	if index != len(s) {
		return
	}

	// value can be empty, so no need to do the check for valueStart != valueEnd
	ok = true
	p.value = s[valueStart:valueEnd]
	return
}
```

Not sure if I missed something in this implementation, but afaik this is following the same pattern as the regex, but let me reiterate, if you prefer your version happy to use that one.

But, if that is the case, would you mind doing those changes yourself. I wouldn't want have to address future comments in a piece of code I didn't write. wdyt?

Comment (pellared) [reply baggage/baggage.go:607]: Take notice that you should use the specification as reference. Not the regex. There is no regex in the specification.

If other @open-telemetry/go-approvers will accept your implementation, I will not oppose. I can always create a separate PR with my proposal. I can totally agree that my suggestion is subjective.

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:555]: Can we set `p.key` only if a valid property is returned (when `ok == true`)?

File: baggage/baggage.go
Comment (pellared) [reply baggage/baggage.go:607]: Here is a similar example PR which has comments referring to the W3C specification: https://github.com/open-telemetry/opentelemetry-go/pull/4722

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go]: Of course I can, I didn't do it because it would just make the code more verbose, but will send the diff.

File: baggage/baggage.go
Comment (pellared) [reply baggage/baggage.go]: I would still appreciate if you try to add some comments and references to the actual specification.

Something like (based on your previous comment where you referenced the regex):

```go
// parsePropertyInternal parses the property defined in https://www.w3.org/TR/baggage/#definition.
func parsePropertyInternal(s string) (p Property, ok bool) {
        // leading key OWS
	index := skipSpace(s, 0)

        // key as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6.
	keyStart := index
	keyEnd := index
	for _, c := range s[keyStart:] {
		if !validateKeyChar(c) {
			break
		}
		keyEnd++
	}

        // empty key
	if keyStart == keyEnd {
		return
	}

        // trailing key OWS
	index = skipSpace(s, keyEnd)

	p.key = s[keyStart:keyEnd]

	// this matches only the key
        // secondary case: key OWS
	if index == len(s) {
		ok = true
		return
	}

        // now let's move to the primary case: 
        // key OWS "=" OWS value
        // this has the same pattern for the key, so we just need to look now for '='
	// now let see if it matches the key and the value
	if s[index] != keyValueDelimiter[0] {
                 // doesn't match the expect char, ok is false, so invalid property
		return
	}

	// we set it to true as soon as we find the delimiter even if it is empty.
        // we don't care to set it as true if the value is invalid, because the caller must check first the boolean return value
	p.hasValue = true

        // leading value OWS
	index = skipSpace(s, index+1)

        // value (baggage-octet) as defined in https://www.w3.org/TR/baggage/#definition.
	valueStart := index
	valueEnd := index
	for _, c := range s[valueStart:] {
		if !validateValueChar(c) {
			break
		}
		valueEnd++
	}

        // trailing value OWS
	index = skipSpace(s, valueEnd)

        // if after checking for trailing whitespaces for the value, we have not finished, it means there are invalid characters, so ok is still false
	if index != len(s) {
		return
	}

	// value can be empty, so no need to do the check for valueStart != valueEnd
	ok = true
	p.value = s[valueStart:valueEnd]
	return
}
```

Regarding my suggestion. I can make a separate PR after this one is merged 😉

Comment (cdvr1993) [reply baggage/baggage.go]: Ahh thanks. I think that change got lost by mistake. Thanks for pointing that out.

**Review by MadVikingGod — APPROVED** [review]

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go:603]: only copying the definition from the regex. `\s` includes all those values, would you prefer if I remove any of those?

File: baggage/baggage.go
Comment (cdvr1993) [reply baggage/baggage.go:563]: because of this one: `strings.Cut(keyValue, keyValueDelimiter)`. We would need to create here a string every time it gets called. Not entirely sure about all the compiler optimizations, but assuming there aren't any. It is cheaper to keep this as a string and use index0 than to create a string every time, wdyt?

File: baggage/baggage.go
Comment (MadVikingGod) [reply baggage/baggage.go:603]: See if that breaks any of our tests. If all tests still pass, then I think we should change it.  If they break I would keep the same behavior.

Comment (cdvr1993) [reply baggage/baggage.go]: it seems it works

**Comment by cdvr1993** [comment]
hi everyone, just checking in, is there anything else missing to merge this? Also, I couldn't find it, what is the process for releasing? Do you release on a schedule or is it on-demand?

**Review by pellared — APPROVED** [review]

**Pull request merged**
